Bookmarks: add documentation (#308)
BSI-Business-Systems-Integration-AG/org.eclipse.scout.docs: docs/modules/technical-guide/pages/user-interface/desktop.adoc

----

By default, this creates a bookmark for the currently selected page.
A different page can be specified via the optional argument `CreateBookmarkParam`.
The same argument allows configuring the type and content of the resulting bookmark.

The return value is a xref:technical-guide:working-with-data/data-object.adoc[data object] containing mainly the bookmark definition itself (={nbsp}the page state and the path).
Depending on the implementation, additional metadata may be included, e.g. a human-readable title.

For a page to be bookmarkable, it needs to have a xref:user-interface/object-identifiers.adoc[unique identifier], so it can be recognized later.
This identifier is usually assigned by specifying the `uuid` property in the page model.
Because all child pages of a table page are of the same type and therefore share the same `uuid`, some additional information is needed to unambiguously identify them:

* If the table has columns marked as _primary key_, the corresponding row values are converted to a `BookmarkTableRowIdentifierDo` and stored in the bookmark.
* If present, the _page param_ of the child page is included in the bookmark.

> This data object contains the necessary input to re-create a specific page instance, e.g. the key of the corresponding entity.

Reviewer: ```suggestion
This data object contains the necessary input to re-create a specific page instance, e.g. the key of the corresponding entity, and needs to have the same name as the page with the suffix `PageParamDo`.
```

> For pages without an explicit page param, a `PageIdDummyPageParam` with the page's `uuid` is stored in the bookmark.

Reviewer: ```suggestion
For pages without an explicit page param, a `PageIdDummyPageParamDo` with the page's `uuid` is stored in the bookmark.
```

=== Activating a Bookmark

Given a `BookmarkDo`, you can instruct the `BookmarkSupport` to navigate to the corresponding location.

[source,javascript]
----
await BookmarkSupport.get(session).activateBookmark(bookmark);
----

This resolves the elements of the page path step-by-step.
Each page is loaded and restored to the previous state, before a drill-down with the next element of the path is performed.
In case one element of the page path cannot be found anymore (e.g. because the corresponding data has been deleted in the meantime), the process stops and a warning is shown to the user.
Otherwise, the last page in the page is selected.

Instead of jumping to the original location, it is also possible to simply create a new instance of a bookmarked page and add it dynamically to the outline.
In that case, the rest of the page path in the bookmark is irrelevant.

> To do so, use the `PageResolver` to compute the corresponding object type from the stored page param.

Reviewer: ```suggestion
To do so, ensure the page has an explicit _page param_ if it requires some input, so it can be re-created without its parent.
With the stored page param in hand, you can compute the corresponding object type of the page by using the `PageResolver`.
```
